daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Equality.scala

```diff
@@ -1,75 +1,103 @@
+          push(
+            new Interlace(xMap.keys.iterator, xMap.values.iterator),
+            new Interlace(yMap.keys.iterator, yMap.values.iterator),
+          )
+        case (SStruct(_, xs), SStruct(_, ys)) =>
+          push(xs.iterator().asScala, ys.iterator().asScala)
+        case (SAny(xType, x), SAny(yType, y)) =>
+          push(Iterator(x), Iterator(y))
+          success = xType == yType
+        case (STypeRep(xType), STypeRep(yType)) =>
+          success = xType == yType
+        case _ =>
+          throw SErrorCrash("try to compare incomparable type")
+      }
 
-  @tailrec
-  private[this] def equality(stack0: FrontStack[(SValue, SValue)]): Boolean = {
-    stack0.pop match {
-      case Some((tuple, stack)) =>
-        tuple match {
-          case (x: SPrimLit, y: SPrimLit) =>
-            x == y && equality(stack)
-          case (SEnum(tyCon1, _, rank1), SEnum(tyCon2, _, rank2)) =>
-            tyCon1 == tyCon2 && rank1 == rank2 && equality(stack)
-          case (SRecord(tyCon1, fields1, args1), SRecord(tyCon2, fields2, args2)) =>
-            tyCon1 == tyCon2 && (fields1 sameElements fields2) &&
-              equality(zipAndPush(args1.iterator().asScala, args2.iterator().asScala, stack))
-          case (SVariant(tyCon1, _, rank1, arg1), SVariant(tyCon2, _, rank2, arg2)) =>
-            tyCon1 == tyCon2 && rank1 == rank2 && equality((arg1, arg2) +: stack)
-          case (SList(lst1), SList(lst2)) =>
-            lst1.length == lst2.length &&
-              equality(zipAndPush(lst1.iterator, lst2.iterator, stack))
-          case (SOptional(None), SOptional(None)) =>
-            equality(stack)
-          case (SOptional(Some(v1)), SOptional(Some(v2))) =>
-            equality((v1, v2) +: stack)
-          case (STextMap(map1), STextMap(map2)) =>
-            map1.keySet == map2.keySet && {
-              val keys = map1.keys
-              equality(zipAndPush(keys.iterator.map(map1), keys.iterator.map(map2), stack))
-            }
-          case (SGenMap(map1), SGenMap(map2)) =>
-            map1.keySet == map2.keySet && {
-              val keys = map1.keys
-              equality(zipAndPush(keys.iterator.map(map1), keys.iterator.map(map2), stack))
-            }
-          case (SStruct(fields1, args1), SStruct(fields2, args2)) =>
-            (fields1 sameElements fields2) && equality(
-              zipAndPush(args1.iterator().asScala, args2.iterator().asScala, stack),
-            )
-          case (SAny(t1, v1), SAny(t2, v2)) =>
-            t1 == t2 && equality((v1, v2) +: stack)
-          case (STypeRep(t1), STypeRep(t2)) =>
-            t1 == t2 && equality(stack)
-          case _ =>
-            false
-        }
-      case _ =>
-        true
+    while (success && stackX.nonEmpty) {
+      if (stackX.head.hasNext != stackY.head.hasNext)
+        success = false
+      else if (stackX.head.hasNext)
+        step((stackX.head.next(), stackY.head.next()))
+      else
+        pop()
```

Comment:
I would find the following a bit easier to understand but I assume the Scala compiler doesn't compile this into particularly efficient code. Or does it?
```suggestion
      (stackX.head.hasNext, stackY.head.hasNext) match {
        case (true, true) => step((stackX.head.next(), stackY.head.next()))
        case (false, false) => pop()
        case _ => success = false
      }
```

Reply:
The Scala compiler is probably not so clever.
I change it anyway.